spring-integration-ip/src/main/java/org/springframework/integration/ip/config/TcpConnectionFactoryFactoryBean.java

```diff
@@ -128,8 +137,11 @@ public final void setBeanFactory(BeanFactory beanFactory) {
 
 	@Override
 	public Class<?> getObjectType() {
-		return this.connectionFactory != null ? this.connectionFactory.getClass()
-											  : AbstractConnectionFactory.class;
+		return this.connectionFactory != null ? this.connectionFactory.getClass() :
+			this.type == null ? AbstractConnectionFactory.class :
+				isServer() ? AbstractServerConnectionFactory.class :
+					isClient() ? AbstractClientConnectionFactory.class :
+						AbstractConnectionFactory.class;
```

Comment:
H-m-m.
For the Java Config I'd make it as generic:
````java
public class TcpConnectionFactoryFactoryBean<F> extends AbstractFactoryBean<F extends AbstractConnectionFactory>
...
private final Class<F> type;

public TcpConnectionFactoryFactoryBean() {
    this.type = (Class<F>) ResolvableType.forClass(this.getClass()).as(EndpointSpec.class).resolveGenerics()[1];
}
````

Another ctor for the XML parse can accept the class directly:
````java
public TcpConnectionFactoryFactoryBean(Class<F> type) {
     this.type = type;
}
````
and the `setType(String type)` can be `@Deprecated`.

WDYT?